cli/flox-activations/src/activations.rs

```diff
@@ -129,22 +261,55 @@ fn activations_json_path(
         .join("activations.json"))
 }
 
+/// {flox_runtime_dir}/{path_hash(flox_env)}/{activation_id}
+pub fn activation_state_dir_path(
+    cache_dir: impl AsRef<Path>,
+    flox_env: impl AsRef<Path>,
+    activation_id: Uuid,
+) -> Result<PathBuf, Error> {
+    Ok(flox_runtime_dir(cache_dir)?
+        .join(path_hash(flox_env))
+        .join(activation_id.to_string()))
+}
+
 /// Returns the parsed environment registry file or `None` if it doesn't yet exist.
 ///
-/// The file can be written with [flox_core::serialize_atomically]
-#[allow(unused)]
-fn read_activations_json(path: impl AsRef<Path>) -> Result<Option<Activations>, Error> {
+/// The file can be written with [write_activations_json].
+/// This function acquires a lock on the file,
+/// which should be reused for writing, to avoid TOCTOU issues.
+pub fn read_activations_json(
```

Comment:
Rip

> fun fact, it seems like both `acquire_*_lock` have been using the lockfile incorrectly, as `Lockfile::open` does not actually lock anything. We still need to call `Lockfile::lock` to actually aqcuire a lock...

Rip

Reply:
> with aquiring a lock separately, it is _on us_ to do so before we attempt to read the file. IMO this leaves more room for mistakes possibly creating race conditions with another write. locking the lockfile with the read avoids that to a degree unless you drop the lock before writing.
> 
> This is mirroring rusts `&mut` semantics where, while you have the file open _to eventually write_ (i.e. `&mut` reference to the file), you cant open it for reading.

I feel strongly we should be consistent. I agree the caller can make mistakes if they don't have to acquire the lock. But we should either always let the caller make that decision or never do so. Because if I copy code for how we're handling activations.json to env-registry.json, it could cause a bug.

I feel less strongly that it may be a good idea to let the caller read without locking. In e.g. `start-or-attach`, we don't need to lock while polling. We could also have `lock_and_read_activations_json` and `read_activations_json_no_lock`. But like I said that's a less strong opinion, and if you think for now we always want to lock, that's fine by me, so long as we're consistent